Updated QirRuntime/README.md, build/test instructions (#591)
microsoft/qsharp-runtime: src/QirRuntime/README.md

# The Native QIR Runtime

> This folder contains the Quantum Intermediate Representation (QIR) Runtime project. QIR is a subset of the [LLVM](https://llvm.org/) Intermediate Representation.

Reviewer: You've used the definite article "the" to refer to LLVM IR, but not to QIR; I'd suggest either adding a "the" before "QIR" or dropping the "the" before "LLVM."

> QIR Runtime includes implementation of the

Reviewer: ```suggestion
The QIR runtime includes an implementation of the
```

 [QIR specification](https://github.com/microsoft/qsharp-language/tree/main/Specifications/QIR) and the bridge to

>  compile QIR to be run against the native full state simulator.

Reviewer 2: ```suggestion
 run QIR against the native full state simulator.
```

## Build

> ### Prerequisites

Reviewer: Blank lines are recommended after Markdown headers.
```suggestion
### Prerequisites

```

The QirRuntime project is using CMake (3.17) + Ninja(1.10.0) + Clang++(11.0.0). Other versions of the tools might work
 but haven't been tested. Only x64 architecture is supported.  

> For runing the PowerShell scripts below use the powershell core (`pwsh`),

Reviewer 2: Is this necessary? I believe with a minor tweak to the condition in prerequisites.ps1 the scripts run just fine on the out of the box installation (they did about three weeks ago in any case).
```suggestion
For running the PowerShell scripts below use the powershell core (`pwsh`),
```

Author: After Stefan significantly refactored our build scripts 1 - 2 weeks ago, for me the .ps1 were failing without `pwsh`.

> For runing the PowerShell scripts below use the powershell core (`pwsh`),

Reviewer: Note that from version 7.0, "PowerShell Core" is just "PowerShell." (Similar to that .NET 5 is just ".NET" even though it's the next version of .NET Core.)

```suggestion
For runing the PowerShell scripts below use PowerShell Core or PowerShell (`pwsh`),
```

> e.g. [this one](https://github.com/PowerShell/PowerShell/releases/download/v7.1.2/PowerShell-7.1.2-win-x64.msi), not the inbox PowerShell.

Reviewer: This links directly to a specific version of the Windows 64-bit installer; I'd suggest linking to the PowerShell installation page, both so that we don't need to update the link when new versions are released, and so that users on other operating systems aren't directed to Windows-only links. E.g.: https://docs.microsoft.com/powershell/scripting/install/installing-powershell.

To install prerequisite tools for building the QIR runtime, you can set the `ENABLE_QIRRUNTIME` environment variable to the string `"true"` 

> and run [`prerequisites.ps1`](prerequisites.ps1), or manually install pre-reqs with the steps listed below
> (Windows steps rely on [Chocolatey](https://chocolatey.org/) for installation).

Reviewer: ```suggestion
and run [`prerequisites.ps1`](prerequisites.ps1), or manually install pre-reqs with the steps listed below.
Note that on Windows, this script relies on the [Chocolatey package manager](https://chocolatey.org/), while on macOS, `prerequisites.ps1` relies on the `brew` package manager.
```

A link to homebrew may also be helpful here.

Author: > A link to homebrew may also be helpful here.

I assume it is https://brew.sh

See [https://code.visualstudio.com/docs/remote/wsl] on how to use VS Code with WSL.

> #### Other Prerequisites

Reviewer: ```suggestion
#### Other Prerequisites

```

The build depends on `Microsoft.Quantum.Simulator.Runtime` dynamic library built at a higher level of the directory tree.
To build that library follow the instructions in [`qsharp-runtime/README.md`](../../README.md#building-from-source)
(up to and including the step `Simulation.sln`).

> ### Build Commands

Reviewer: ```suggestion
### Build Commands

```

To build QirRuntime you can run [`build-qir-runtime.ps1`](build-qir-runtime.ps1) script from QirRuntime folder:
```batch
pwsh build-qir-runtime.ps1
```
  
The script will create the `build/{Debug|Release}` folder and place the build artifacts in it. The configuration `Debug|Release`
is specified with the `BUILD_CONFIGURATION` environment variable.
If the variable is not set then the default is specified in [`set-env.ps1`](../../build/set-env.ps1).  
We strongly recommend doing local builds using this [`build-qir-runtime.ps1`](build-qir-runtime.ps1) build script 
because it is used by the continuous integration (CI) infrastructure, and this script also runs clang-tidy if it is installed.

> Or you can use CMake directly (not recommended). For example in Windows you can enter the following commands in the command line:

Reviewer: ```suggestion
Though it is not recommended, you can also use CMake directly. For example, in Windows you can enter the following commands in the command line:
```

```batch

> @rem Navigate to QirRuntime folder.

Reviewer: This looks like it uses cmd notation, but cmd is no longer the default shell on Windows (PowerShell was made the default several years ago).

Author: Yes... I'm out-of-date ;-)
Don't you have on top of your head, what I should use instead of  ```batch
for PowerShell syntax highlighting? Locally I try `ps`, `ps1`, `PowerShell` but my MarkDown plugin for Chrome does not highlight the syntax if I use any of those.

Author: Oh. looks like it is shown [here](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1511).

@rem Creatge the `build` directory:
mkdir build

@rem Enter the `build` directory:
pushd build

@rem Prepare for the build:
cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug ..

@rem Build (you will likely need to issue this command several times, see issue #589):

> cmake --build .

Reviewer 3: I still don't understand this. Why does this need to be run multiple times? I haven't observed that behavior, though I'm trying now on a fresh repo to see if there is something I'm missing.

Author: I saw similar behavior during multithreaded builds ( `make -j ..` ). Different targets are built by different threads. And sometimes the later targets (that depend on the results of the earlier target builds) start earlier than the earlier target build  fully completes. It felt like the earlier target step was completed (from the point of view of the build utility), but the file output of the earlier stage was cached for a few seconds (e.g. the object file was not written to the disk completely), but the linking stage was already kicked off, and the linker was failing to open the object file, and the linking was failing.
At that moment the cached file was flushed onto the disk and the repeated build attempt was succeeding.

Reviewer 3: I haven't observed this behavior with CMake, which does not perform parallel builds by default. In fact, CMake will build in dependency order based on detected dependencies across different targets, with the order of the folders in the top level CMakeLists.txt being the fallback. This ensures that CMake will only need to be run once. Have you gotten this error in your local builds?

I did just try it again, and after running bootstrap.ps1 and copying the relevant .ll files to the correct locations, I was able to manually create `src\QirRuntime\build\Debug` and then run these two commands to successfully build the whole QIR runtime and tests:
```
cmake -G Ninja ../..
cmake --build . --target install
```

Reviewer 3: Also worth noting, the `--target install` ensures that all the binary output is collected in the `bin` folder, which the test script expects.

Author: I just removed those steps.

Author: I double checked out of curiosity. Looks like `--target install` was the missing part.

@rem Return from the `build` directory:
popd
```
## Tests

> ### Running All Tests

Reviewer: ```suggestion
### Running All Tests

```

> ```batch

Reviewer: Similarly here, the `@rem` notation is specific to cmd.exe.

To build that library follow the instructions in [`qsharp-runtime/README.md`](../../README.md#building-from-source)
(up to and including the step `Simulation.sln`).

### Build Commands

To build QirRuntime you can run [`build-qir-runtime.ps1`](build-qir-runtime.ps1) script from QirRuntime folder:
```batch
pwsh build-qir-runtime.ps1
```
  
The script will create the `build/{Debug|Release}` folder and place the build artifacts in it. The configuration `Debug|Release`
is specified with the `BUILD_CONFIGURATION` environment variable.
If the variable is not set then the default is specified in [`set-env.ps1`](../../build/set-env.ps1).  
We strongly recommend doing local builds using this [`build-qir-runtime.ps1`](build-qir-runtime.ps1) build script 
because it is used by the continuous integration (CI) infrastructure, and this script also runs clang-tidy if it is installed.

Though it is not recommended, you can also use CMake directly.
For example, in Windows you can enter the following commands in the command line:
```powershell
# Navigate to QirRuntime folder.

# Creatge the `build` directory:

> mkdir build

Reviewer 3: to match what the script does, the directory should be `build/Debug`, then navigate into that folder.

Author: I removed those steps, they have some more issues.

# Enter the `build` directory:
pushd build

# Prepare for the build:

> cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug ..

Reviewer 3: This needs to be `cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug ../..` since the pwd should be `src\QirRuntime\build\Debug`

Author: I removed those steps, they have some more issues.
